### PR TITLE
add model and nonlinearity

### PR DESCRIPTION
**Pull request by pcrespo2017**
```python
# Jax
import jax
from vintch_model.subunit_model import SubunitModel

jax_model = SubunitModel(n_basis_funcs=200, backend="jax")
key = jax.random.PRNGKey(10)
jax_input = jax.random.normal(key, (1, 1, 16, 16, 16))
jax_output = jax_model(jax_input)

# PyTorch
import torch
from vintch_model.subunit_model import SubunitModel

torch_model = SubunitModel(n_basis_funcs=200, backend="torch")
torch_input = torch.randn(1, 1, 16, 16, 16)
torch_output = torch_model(torch_input)
```

### CODE REVIEW

File: src/vintch_model/subunit_model.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        self.kernel_time_depth = kernel_time_depth
+        self.n_dim = n_dim
+        self.n_channels = n_channels
+        self.backend_class = get_backend(backend)
+        self._kernels = self.backend_class.randn(
+            (n_channels, 1, kernel_time_depth, kernel_size, kernel_size)
+        )
+        if pooling_in_time:
+            self._pooling_weights = self.backend_class.randn(
+                (n_channels, n_dim, n_dim, n_dim)
```
Comment (billbrod) [inline src/vintch_model/subunit_model.py:61]: I think this shape should be constructed similarly to what I said about kernel size about: have a single `pooling_size` argument, which is a tuple of ints of arbitrary shape. that also allows us to get rid of the pooling_in_time argument: if the length of this input is the same as `kernel_size`, then we're pooling in time. If not, it must be `len(kernel_size)-1` and we're not pooling in time

File: src/vintch_model/subunit_model.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        self.backend_class.check_input_type(biases)
+        self._pooling_biases = biases
+
+    def forward(self, x):
+        """
+        Forward pass through the whole model.
+        """
+        self.backend_class.check_input_type(x)
+        # ensure the input is 5D and the second dimension is 1 (grayscale)
+        if x.ndim != 5 or x.shape[1] != 1:
```
Comment (billbrod) [inline src/vintch_model/subunit_model.py:130]: @BalzaniEdoardo is there a way to avoid an if statement here while still validating the shape? It will fail at the call to convolution if the input is not 5d (so we can use a try/except and raise a more informative error message there), but I'm not sure about ensuring the channel dimension is 1 (which we want because we only want to deal with grayscale, at least for now)

File: src/vintch_model/subunit_model.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            raise ValueError(
+                f"Input must be a 5D tensor with shape [batch, {self.n_channels}, time, height, width], got {x.shape} instead."
+            )
+        generator_signal = self.generator_forward(x)  # [batch, time]
+        out = self.nonlinearity_out.transform(generator_signal)  # [batch, time]
+        return out
+
+    def generator_forward(self, x):
```
Comment (billbrod) [inline src/vintch_model/subunit_model.py:138]: generator is an odd name. I know they call it "the generator signal" in the paper, but [generator](https://wiki.python.org/moin/Generators) means something specific in python. so maybe call it "generator signal" instead?

and add a docstring here which explains this is gets sent through the final nonlinearity in order to be turned into the cell's firing rate.

File: src/vintch_model/subunit_model.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        pooled = self.weighted_pooling(
+            activated, self.pooling_weights, biases=self.pooling_biases
+        )  # [batch, channels, time]
+        output = pooled.sum(axis=1)  # [batch, time]
+        return output
+
+    def convolve(self, x, kernel):
+        return self.backend_class.convolve(x, kernel)
+
+    def nonlinearity_pass(self, x, nonlinearities):
```
Comment (billbrod) [inline src/vintch_model/subunit_model.py:154]: I'm confused by this name, why "pass"?

File: src/vintch_model/subunit_model.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        transformed = [
+            nonlinearity.transform(x[:, c])
+            for c, nonlinearity in enumerate(nonlinearities)
+        ]
+        return self.backend_class.lib.stack(transformed, axis=1)
```
Comment (billbrod) [inline src/vintch_model/subunit_model.py:159]: I think stack can also be handled by [einops.pack](https://einops.rocks/4-pack-and-unpack/#stacking-data-layers)

Comment (billbrod) [reply src/vintch_model/subunit_model.py:159]: which I'd prefer

File: src/vintch_model/subunit_model.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                "batch_size gray_channel time height width, n_channels time height width -> batch_size n_channels time",
+            )
+        else:
+            pooled = einsum(
+                x,
+                pooling_weights,
+                "batch_size gray_channel time height width, n_channels height width -> batch_size n_channels time",
+            )
+
+        if biases is not None:
```
Comment (billbrod) [inline src/vintch_model/subunit_model.py:175]: why would biases be None?

**Review by billbrod — CHANGES_REQUESTED** [review]
Lot of small comments here, let me know if it would be helpful to talk through any of these!

- In `__init__.py`, add
  ```python
  from . import nonlinearity
  from . import subunit_model
  ```
  that way, users can call `vintch_model.subunit_model.SubunitModel` and `vintch_model.nonlinearity` (I don't think we need to expose the backend classes)
- I think we'll want to expose the convolution padding, make that an argument the user can set
- when we discussed, we talked about the "soft" vs. "hard" tent nonlinearity -- I would refer to "soft" as "tents" and "hard" as either "one-hot" or "delta", @BalzaniEdoardo what do you think ?
  - but also, do we need both? don't they just use the soft / tent in the paper?
- for the nonlinearity, I see that you are matching the nemos basis API. talking with Edoardo, we don't think that's necessary: this basis is only used for getting the nonlinearity in this model, and we won't people to swap it out for other ones. so it should really only have a `__call__` method (and maybe forward, to match torch's API)
  - once you've made that change, it will be easier for me to look through how it works  -- I'm not understanding what exactly is happening there.

**Review by BalzaniEdoardo — CHANGES_REQUESTED** [review]
For this round, I focused on the backend side and I think it is on the right track. 

Unit testing for checking that our convolve and all the function we are using results in the same I/O independently from the backend is crucial. it is better to have these tests sooner rather than later.

And we should have end-to-end test of the type: apply the forward pass with `jax` and then with `torch` and match outputs. It think this are safeguards that we should have in place from early on.

We should use try different dtypes and shapes for the arrays, because the two packages may have different promotion behaviors (for example when you convolve int arrays or int and float arrays and stuff like that) and also different broadcasting.

File: src/vintch_model/backend_config.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    def __init__(self):
+        import torch
+
+        self.lib = torch
+        self.randn = torch.randn
+        self.conv3d = torch.nn.functional.conv3d
+        self.name = "torch"
+
+    def atleast_1d(self, x):
+        return x if x.dim() > 0 else x.unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/backend_config.py:14]: I can't think how? which is the issue with the if statement?

File: src/vintch_model/backend_config.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
+import numpy as np
+
+
+class TorchBackend:
+    def __init__(self):
+        import torch
+
+        self.lib = torch
+        self.randn = torch.randn
+        self.conv3d = torch.nn.functional.conv3d
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/backend_config.py:10]: you're right, removed

File: src/vintch_model/backend_config.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+class JaxBackend:
+    def __init__(self):
+        import jax.numpy as jnp
+        from jax import random, lax
+
+        self.lib = jnp
+        self.random = random
+        self.atleast_1d = jnp.atleast_1d
+        self.lax = lax
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/backend_config.py:36]: lax for example is useful since self.lib here is jax numpy but lax is part of jax. We could set self.lib = jax, and write the rest as jax.numpy.'method'

File: src/vintch_model/nonlinearity.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        x = self.backend.atleast_1d(x)
+        original_shape = x.shape
+        flat_x = x.flatten()
+
+        features = self.evaluate(flat_x)
+        output_flat = self.apply_weights(features)
+        return output_flat.reshape(original_shape)
+
+
+def tent_nonlinearity_jax_to_torch(jax_instance):
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/nonlinearity.py:154]: at least for now, I think it is useful to have functions to convert the model from jax to torch. What do you think?

File: src/vintch_model/subunit_model.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+    def forward(self, x):
+        """
+        Forward pass through the whole model.
+        """
+        self.backend_class.check_input_type(x)
+        # ensure the input is 5D and the second dimension is 1 (grayscale)
+        if x.ndim != 5 or x.shape[1] != 1:
+            raise ValueError(
+                f"Input must be a 5D tensor with shape [batch, {self.n_channels}, time, height, width], got {x.shape} instead."
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/subunit_model.py:132]: you're right

File: src/vintch_model/subunit_model.py (hunk 4 above)
Comment (pcrespo2017) [reply src/vintch_model/subunit_model.py:154]: there is no reason why, do you have an alternative name?

File: src/vintch_model/subunit_model.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+            pooled = einsum(
+                x,
+                pooling_weights,
+                "batch_size gray_channel time height width, n_channels time height width -> batch_size n_channels time",
+            )
+        else:
+            pooled = einsum(
+                x,
+                pooling_weights,
+                "batch_size gray_channel time height width, n_channels height width -> batch_size n_channels time",
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/subunit_model.py:172]: we shouldn't add time here, right?

File: src/vintch_model/backend_config.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
+import numpy as np
+
+
+class TorchBackend:
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/backend_config.py:4]: I don't understand this

File: src/vintch_model/backend_config.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    Backend class for NumPy operations.
+
+    This class provides methods for NumPy array operations.
+    """
+
+    def __init__(self):
+        self.lib = np
+        self.name = "numpy"
+
+    def convolve(self, x, kernel, padding="same"):  # MISSING IMPLEMENTATION
```
[the remark this answers is not in the document]
Comment (billbrod) [reply src/vintch_model/backend_config.py:187]: also, why isn't it implemented?

File: src/vintch_model/nonlinearity.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        n_basis_funcs: int = 15,
+        nonlinearity_type: Literal["relu", "quadratic"] = "relu",
+    ):
+        if n_basis_funcs < 1:
+            raise ValueError("basis functions must be a positive integer.")
+
+        self._n_basis_funcs = n_basis_funcs
+        self._backend = backend_instance
+
+        self._build_target_output = (
```
[the remark this answers is not in the document]
Comment (billbrod) [reply src/vintch_model/nonlinearity.py:35]: by "this", I mean I don't know what "target output" means

File: src/vintch_model/nonlinearity.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+        original_shape = x.shape
+
+        features = self._tents_transform(x)
+
+        output_flat = einops.einsum(
+            features, self._weights, "extended_dim n_basis, n_basis -> extended_dim"
+        )
+
+        return output_flat.reshape(original_shape)
```
[the remark this answers is not in the document]
Comment (billbrod) [reply src/vintch_model/nonlinearity.py:151]: could add a test for this -- pass a weird-shaped input through this and then apply the same nonlinearity in some other way (e.g., relu(x) = `x.clip(min=0)`) and check that all the values are the same

File: tests/test_subunit_model.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        )
+
+        output = model(x_backend)
+        outputs.append(output)
+
+    output_shapes = [o.shape for o in outputs]
+    assert all(
+        shape == output_shapes[0] for shape in output_shapes
+    ), "Output shapes do not match across backends."
+    assert np.allclose(
```
Comment (billbrod) [inline tests/test_subunit_model.py:57]: when I run this function locally setting dtype=np.float64, I end up with `output[1]` is float32, which is a bug. so I'd have a test that checks when dtype is set you get the expected output dtype, and also why that's not happening for jax

**Review by billbrod — CHANGES_REQUESTED** [review]
This is looking good, most of my comments are more finicky now, the overall structure of this is good!

File: src/vintch_model/nonlinearity.py [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+        Get the backend instance.
+        """
+        return self._backend
+
+    @backend.setter
+    def backend(self, backend_instance: BackendBase):
+        self._backend = backend_instance
+
+    @property
+    def grid(self) -> Tensor:
```
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/nonlinearity.py:72]: grid is just the array of functions centers, maybe something like `fun_centers` is a better name? also it doesn't need to be public now, I will remove it

File: src/vintch_model/nonlinearity.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (pcrespo2017) [reply src/vintch_model/nonlinearity.py:35]: well, this is going away. I thought we talked about using least squares to getting the initial weights but it's actually not needed.

File: src/vintch_model/nonlinearity.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        self._n_basis_funcs = n_basis_funcs
+        self._backend = backend_instance
+
+        self._initialize_weights = (
+            self._initialize_relu_weights
+            if nonlinearity_mode == "relu"
+            else self._initialize_quadratic_weights
+        )
+        self._initialize_basis_functions()
+        self._initialize_weights()
```
Comment (billbrod) [inline src/vintch_model/nonlinearity.py:46]: Given that the initialize weights method is never called again, I'd rearrange this. what you have is confusing to me -- I thought the initialize weights methods were being called before `_initialize_basis_functions` and I couldn't figure out how that was possible. 
```suggestion
        self._initialize_basis_functions()
        if nonlinearity_mode == "relu":
            self._initialize_relu_weights()
        elif nonlinearity_mode == "quadratic":
            self._initialize_quadratic_weights()
```

File: tests/test_nonlinearity.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+        non_lin_output = non_linearity(x)
+
+        if nonlinearity_mode == "relu":
+            expected_output = x_np.clip(min=0)
+        elif nonlinearity_mode == "quadratic":
+            expected_output = x_np**2
+
+        assert backend.to_numpy(non_lin_output) == pytest.approx(
+            expected_output, abs=1e-5
+        )
```
Comment (billbrod) [inline tests/test_nonlinearity.py:40]: I would also explicitly test that the shapes are the same. that's probably happening, but it's worth being explicit.

File: tests/test_nonlinearity.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    @pytest.mark.parametrize(
+        "nonlinearity_mode",
+        ["relu", "quadratic"],
+    )
+    def test_reshapes(self, backend_cls, nonlinearity_mode):
+        backend = backend_cls()
+        non_linearity = TentNonlinearity(
+            backend_instance=backend, nonlinearity_mode=nonlinearity_mode
+        )
+        x = non_linearity._tent_centers
```
Comment (billbrod) [inline tests/test_nonlinearity.py:28]: why is this the input? nonlinearity should work with arbitrarily shaped inputs, right? I'd try many different shapes.

File: src/vintch_model/nonlinearity.py [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+    Tent-based nonlinearity.
+
+    Parameters
+    ----------
+    n_basis_funcs :
+        Number of basis functions to use. Must be a positive integer.
+    backend_instance :
+        Backend to use. Must be an instance of BackendBase.
+    nonlinearity_mode :
+        Type of nonlinearity to initialize weights for. Weights will be fitted to data.
```
Comment (billbrod) [inline src/vintch_model/nonlinearity.py:20]:
```suggestion
        Type of nonlinearity to initialize weights for. 
```

**Review by billbrod — CHANGES_REQUESTED** [review]
Just a couple of very small changes and then this is ready to go!

File: src/vintch_model/subunit_model.py [hunk 23]
```diff
@@ -44,30 +44,29 @@ def __init__(
         self._kernels = self._backend.randn((n_channels, 1, *subunit_kernel))
```
Comment (pcrespo2017) [inline src/vintch_model/subunit_model.py:44]: should check the shape of input as if it is 2D and kernel 3D it gives NaN outputs

**Review by billbrod — APPROVED** [review]
Just make the small addition to the docstring I mentioned, and this is good to merge! I don't need to review again.
